system/lib/wasmfs/syscalls.cpp

```diff
@@ -1121,4 +1129,65 @@ long __syscall_ftruncate64(long fd, long low, long high) {
   }
   return ret;
 }
+
+long __syscall_pipe(long fd) {
+  auto* fds = (__wasi_fd_t*)fd;
+
+  // Make a pipe: Two PipeFiles that share a single data source between them, so
+  // that writing to one can be read in the other.
+  //
+  // No backend is needed here, so pass in nullptr for that.
+  auto data = std::make_shared<PipeData>();
+  auto reader = std::make_shared<PipeFile>(S_IRUGO, data);
+  auto writer = std::make_shared<PipeFile>(S_IWUGO, data);
+
+  auto fileTable = wasmFS.getFileTable().locked();
+  fds[0] =
+    fileTable.addEntry(std::make_shared<OpenFileState>(0, O_RDONLY, reader));
+  fds[1] =
+    fileTable.addEntry(std::make_shared<OpenFileState>(0, O_WRONLY, writer));
+
+  return 0;
+}
+
+int __syscall_poll(struct pollfd* fds, nfds_t nfds, int timeout) {
+  auto fileTable = wasmFS.getFileTable().locked();
+
+  // Process the list of FDs and compute their revents masks. Count the number
+  // of nonzero such masks, which is our return value.
+  long nonzero = 0;
+  for (nfds_t i = 0; i < nfds; i++) {
+    auto* pollfd = &fds[i];
+    auto fd = pollfd->fd;
+    if (fd < 0) {
+      // Negative FDs are ignored in poll().
+      continue;
+    }
+    // Assume invalid, unless there is an open file.
+    auto mask = POLLNVAL;
+    auto openFile = fileTable.getEntry(fd);
+    if (openFile) {
+      mask = 0;
+      auto flags = openFile->getFlags();
+      if (flags == O_WRONLY || flags == O_RDWR) {
+        mask = POLLOUT;
+      }
+      if (flags == O_RDONLY || flags == O_RDWR) {
+        // If there is data in the file, then there is also the ability to read.
+        if (openFile->locked().getFile()->locked().getSize() > 0) {
```

Comment:
Does this also need to check that the current position is not at the end of the file, where there is no additional data to read?

Reply:
Very good question. I re-read the docs and I'm still not sure. I'll add a TODO here to consider. We may also find we have a test later on that clarifies this when we get to it.

Reply:
TODO sgtm.